README.md

- For instructions, visit [https://inteltechniques.com/voip.suite.html](https://inteltechniques.com/voip.suite.html)
- Or the [Wiki](https://github.com/0perationPrivacy/voip/wiki) section.

- View the [Changelog](CHANGELOG.md)

- Report [Security Issues](SECURITY.md)

- [Troubleshooting](https://github.com/0perationPrivacy/VoIP/wiki/Troubleshooting) tips

- Donate: https://www.OperationPrivacy.com/donate
  - **BTC (Bitcoin):** bc1qh592gld4u2m8vkedxr558l82066jwj5mgvspzq
  - **XMR (Monero):** 899MAgiqJpx6p2gKiizokTKY18o1arRddLXmojVt55A1caQZrHW24AiZoZ2zNJJtr7RAct8cXiwNbUcg9HMXtwTsHJCyXUt

![VoIP-Suite](https://user-images.githubusercontent.com/89225730/134289428-da8bcf88-5145-4ed6-9d6b-70d57b084e8f.png)


![VoIP-Suite-Overview](https://user-images.githubusercontent.com/89225730/204443981-d89391f3-b832-4f55-840e-32c665d9b4c1.png)


![GitHub forks](https://img.shields.io/github/forks/0perationPrivacy/voip?style=flat-square)
![GitHub](https://img.shields.io/github/license/0perationPrivacy/voip?style=flat-square)


CHANGELOG.md
# Changelog

> Note: Every update deletes the MMS media files and folder which is stored locally in your ephemeral cloud hosted service and not stored in the database.

---
## v0.94 (Jul 24, 2023)
### Feature
- added autocomplete/search in dropdown under `Profile Settings` > `Get Number` to search through a long list of numbers

## v0.93 (Jul 14, 2023)
### Security
- stop sending MFA secret to client upon login
  - doing so would allow anyone to generate the TOTP after logging in with user/pass
### Bug Fixes
- fixed chat conversations not appearing (Issue #204)

## v0.92 (Jul 13, 2023)
### Bug Fixes
- fixed a few leading slash issues with a helper function
- fixed socket.io-client import from throwing errors
- fixed build errors

## v0.91 (Jul 11, 2023)
### Feature
- Ability to send PGP Encrypted email notifications of smses received
### Bug Fixes
- fixed a few leading slash issues
- fixed hitting enter on OTP
- fixed path url 404 not found on new registrations
- added example in .env for custom application path for url obscurity purposes

## v0.90 (Jan 2, 2022)
### Feature
- backend sockets implementation for testing the mobile app
### Bug
- attachments file extension fixed

## v0.89 (April 1, 2022)
### Bug
- App would randomly crash with some API pulls, hard to replicate. Added a lot of try/catch statements to contain it.

## v0.88 (March 20, 2022)
### Typo
- Email settings: TO, FROM field labels corrected.

## v0.87 (Feb 16, 2022)
### Bug
- Messages are removed from Twilio's internal logs with an API call. Sometimes it was failing to delete. Now message log deletion waits for confirmation from the API and retries 5 times if one is not received before giving up. Fixes issue #112
- Fixed issue #105 (images in mobile view were saving as .bin)
- Fixed version display
- TOTP toggle asks for confirmation before disabling

## v0.86 (Feb 14, 2022)
### Bug
- Fixed issue with web browser based calling on a Twilio number.

## v0.85 (Feb 12, 2022)
### Feature
- Hardware Key support added. You can add multiple security keys that support U2F (e.g., Yubikey, OnlyKey, Titan, etc).

## v0.84 (Jan 16, 2022)
### Bug
- fixed issue #107, #110, email settings save issue, custom APPDIRECTORY caching issue.

## v0.83 (Jan 12, 2022)
### Bug
- fixed issue 107, email settings save issue.

## v0.82 (Jan 3, 2022)
### Bug
- fixed issue 103, inbound call option stays open even after call hangs up.

## v0.81 (Dec 9, 2021)
### Bug
- Custom APPDIRECTORY route issue fixed

## v0.80 (Nov 24, 2021)
### Security
- Headers refined, directory exposure reduced
- Initial code added for Hardware Keys (function, frontend, library, conditions, db) (final testing remaining for the next release)

### UI
- dial pad colors, alignment adjusted
- Minor bug fixes

- Open Source License added (GPLv3)

## v0.79 (Nov 13, 2021)
### Bugs
- Bug fixes
### Features
- Add dial pad and DTMF tones

## v0.78 (Nov 12, 2021)
### Security
- **MFA - TOTP** QR code (or manual entry) functionality added!
- Few more syntax edits for input sanitization (prevent NOSQL injections)

## v0.77 (Nov 11, 2021)
### UI/Bug
- Added cancel button on attachment window

### Security
- Added Secure Cookie session values
- Trust Proxy
- NOSQL injection protection of functions, input sanitization (100+ syntax edits)
- Secure functions (Math. to crypto.)
- Added compression
- Expose only directories needed, hiding root

## v0.76 (Nov 9, 2021)
### UI
- Text message overflow word break for very long links fixed

## v0.75 (Nov 9, 2021)
### Security
- Protections added as security headers for: XSS, NoSniff, HSTS, X-Powered-By, FrameGuard, DNS Prefetch, Content Security Policy.

### Bugs
- Pull down to refresh now updates the notification dot and the profile inside the dropdown
- Redirect issue to 404 fixed for non-custom directory


## v0.74 (Nov 2, 2021)
- **UI:** Error page update


## v0.73 (Nov 2, 2021)
### Security
- ***Security through Obscurity***: Put in a random directory name in `.env` under `APPDIRECTORY`. Your application will be served only on that subdirectory.
People will not be able to browse to your app page (by discovering your public github page and _deployment_ link)
_Example:_
>|||
>|---|---|
>|APPDIRECTORY|789gh8ag96lgw7ag8fghlkg|
>|||
By default it is not defined. Full instructions [here](https://github.com/0perationPrivacy/voip/wiki/Bonus-Steps).
- ### Bug fix:
  - fixes search bar cursor size on Chromium browsers (Chrome/Brave)
  - Delete account error on empty password fixed
  - SMS reply to shortcode bug fixed. If provider rejects it, you will see their error message.
  - Minor changes in code to function names, variables, comments

- ### UI:
  - Fixed loading bar size and overlapping issues
  - **version** is now viewable under **Settings** and **Login** page (removed from .env)
  - Full message timestamps (M D, Y H:M)
  - On compose message, contacts dropdown now clears after seleting a contact
  




## v0.72 (Oct 30, 2021)
- ***Update ribbon*** shows when there's a newer version available. Links it to the changelog

## v0.71 (Oct 30, 2021)
- Added _plus icon_ for contact not in list
- New **horizontal bar Loader** visible as messages load
- **Fallback URL** added (_redundancy_, can host on 2 servers now with the same login). You will receive the sms/call on both servers.
- Removed error when deleting an already deleted number from Telnyx/Twilio

## v0.70 (Oct 27, 2021)

### Contacts
- Auto sorting contacts alphabetically by first name
- Searching through contacts

### Account Section added
- Change username option added 
- Change password option added
- Delete Account option added (will delete everything)

### Security
- Rate limiter feature added (mitigates brute forcing the app)
- New https variable added (will enable in next update)

### Search
- Search messages
- Search Contacts

### Bugs/Enhancements
- Set min number of phone digits in contacts to 5
- New profile creation switches to that profile now

app.js
const config =  require('./config.js');

const express = require('express')
const app = express()
var bodyParser = require('body-parser')
const cors = require("cors");
require('dotenv').config()
const path = require('path');
var session = require('cookie-session')
var compression = require('compression')
app.use(compression())

var expiryDate = new Date(Date.now() + 60 * 60 * (1000 * 12 * 30)) // 30 day
app.use(session({
  name: 'session',
  keys: [process.env.COOKIE_KEY, process.env.COOKIE_KEY2],
  cookie: {
    secure: true,
    httpOnly: true,
    expires: expiryDate
  }
}))

let setCache = function (req, res, next) {
  const period = 60 * 60 * 24
  if (req.method == 'GET') {
    res.set('Cache-control', `public, max-age=${period}`)
  } else {
    res.set('Cache-control', `no-store`)
  }
  next()
}
app.use(setCache)

var RateLimit = require('express-rate-limit');
const helmet = require("helmet");
app.use(
  helmet.contentSecurityPolicy({
    useDefaults: true,
    reportOnly: false,
    directives: {
      "default-src": ["'self'", "sdk.twilio.com","wss:","ws:","eventgw.twilio.com"
    ],
      "object-src": ["'self'"],
      "script-src": ["'self'","'unsafe-eval'", "'unsafe-inline'"]
    },
  })
);
//sdk.twilio.com
app.use(helmet.dnsPrefetchControl());
app.use(helmet.expectCt());
app.use(helmet.frameguard());
app.use(helmet.hidePoweredBy());
app.use(helmet.hsts());
app.use(helmet.ieNoOpen());
app.use(helmet.noSniff());
app.use(helmet.permittedCrossDomainPolicies());
app.use(helmet.referrerPolicy());
app.use(helmet.xssFilter());
app.disable('x-powered-by');
app.set('trust proxy', 1)
const server = require('http').createServer(app);

global.io = require('socket.io')(server,{ cors: { origin: '*' } });

var mongoose = require('./config/db.config');

const db = mongoose.connection;
db.on('error', console.error.bind(console, 'connection error:'));
db.once('open', function() {
  console.log('database connected successfully!');
});

//app.use(cors());
app.use(cors({ origin: ['http://localhost:8080'], }))

var limiter = new RateLimit({
  windowMs: 1 * 60 * 1000, // 1 minute
  max: 100,
  message: "Slow down your requests!",
  headers: false
});
  
// apply rate limiter to all requests
app.use(limiter);
io.on('connection', socket => {
  console.log('a user connected');
  socket.on('join_channel', (channel) => {
    console.log(`${channel} user joined channel`);
    socket.join(channel);
  });
  socket.on('join_profile_channel', (channel) => {
    console.log(`${channel} user joined channel`);
    socket.join(channel);
  });
});

app.use('/frontend/dist/index.html', express.static('frontend/dist/index.html'));
app.use('/version.md', express.static('version.md'));
// app.enable('trust proxy')
if( process.env.HTTPS.trim() === 'true'){
  app.use((req, res, next) => {
    if (req.header('x-forwarded-proto') !== 'https'){
      if(req.url == '/get-base-url'){
        next()
        // res.status(200).json({url: process.env.BASE_URL.trim()});
      }else{
        res.sendFile(path.join(__dirname, './error/index.html'));
      }
    } else {
      next()
    }
  })

  /* app.use((req, res, next) => {
    console.log(req.url)
    console.log(req.secure)
    if (req.secure || req.url === '/error') {
      next()
    } else if(req.url == '/get-base-url'){
      res.status(200).json({url: process.env.BASE_URL.trim()});
    }else{
      // res.sendFile(path.join(__dirname, './error/index.html'));
    }
  }) */
}
// parse requests of content-type - application/json
app.use(bodyParser.json({limit: '500mb',parameterLimit: 10000000})); 

// parse requests of content-type - application/x-www-form-urlencoded
app.use(bodyParser.urlencoded({ extended: true, limit: '500mb',parameterLimit: 10000000 }));
app.use('/uploads', express.static('uploads'));
app.use('/src', express.static('src'));
app.use('/frontend', express.static('frontend'));
// app.use('/frontend', express.static('frontend'));
app.use('/frontend/dist/static/', express.static('frontend/dist/static'));
app.get(`/error`, function (req, res) {
  res.sendFile(path.join(__dirname, './error/index.html'));
})
app.get(`/:id`, function (req, res) {
  res.sendFile(path.join(__dirname, './frontend/dist/index.html'));
})
app.get(`/:id/:name`, function (req, res) {
  res.sendFile(path.join(__dirname, './frontend/dist/index.html'));
})
app.use(express.static(path.join(__dirname, './frontend/dist')));

require("./app/routes/auth.route")(app);
require("./app/routes/setting.route")(app);
require("./app/routes/profile.route")(app);
require("./app/routes/media.route")(app);
require("./app/routes/contact.route")(app);
require("./app/routes/email.route")(app);
require("./app/routes/call.route")(app);
require("./app/routes/hardwarekey.route")(app);


app.get('/api/users/', function (req, res) {
  res.status(200).json({message: 'success'});
})

app.get('/get-base-url', function(req, res) {
  res.status(200).json({url: process.env.BASE_URL.trim()});
});
server.listen(process.env.PORT)